enterprise/tools/upload_local_snapshot/upload_local_snapshot.go

```diff
@@ -0,0 +1,232 @@
+	}
+	defer os.RemoveAll(tmpDir)
+
+	// Add digests for regular snapshot files
+	for _, file := range localACResult.OutputFiles {
+		digests = append(digests, file.GetDigest())
+	}
+
+	// Add digests for chunked snapshot files
+	for _, chunkedFileMetadata := range localACResult.OutputDirectories {
+		// Add digests for chunked file metadata objects
+		digests = append(digests, chunkedFileMetadata.GetTreeDigest())
+
+		// Add digests for chunked file chunks
+		tree, err := loader.ChunkedFileTree(ctx, remoteInstanceName, chunkedFileMetadata, tmpDir)
+		if err != nil {
+			log.Fatalf("Failed to process chunked file tree: %s", err)
+		}
+		for _, chunk := range tree.GetRoot().GetFiles() {
+			digests = append(digests, chunk.GetDigest())
+		}
+	}
+	return digests
+}
+
+func getMissingDigestsRemoteCache(ctx context.Context, env environment.Env, remoteInstanceName string, digests []*repb.Digest) []*repb.Digest {
+	req := &repb.FindMissingBlobsRequest{
+		InstanceName:   remoteInstanceName,
+		BlobDigests:    digests,
+		DigestFunction: repb.DigestFunction_BLAKE3,
+	}
+	res, err := env.GetContentAddressableStorageClient().FindMissingBlobs(ctx, req)
+	if err != nil {
+		log.Fatalf("Error FindMissingBlobs: %s", err.Error())
+	}
+	return res.MissingBlobDigests
+}
+
+func uploadDigestsRemoteCache(ctx context.Context, env environment.Env, remoteInstanceName string, digests []*repb.Digest) {
+	tmpDir, err := os.MkdirTemp("", "upload-local-ss-*")
+	if err != nil {
+		log.Fatalf("Error making temp dir: %s", err.Error())
+	}
+	defer os.RemoveAll(tmpDir)
+
+	for _, d := range digests {
+		upload := func() {
+			path := filepath.Join(tmpDir, d.GetHash())
+			node := &repb.FileNode{
+				Digest: d,
+			}
+			inFC := env.GetFileCache().FastLinkFile(ctx, node, path)
+			if !inFC {
+				log.Fatalf("Digest %s does not exist in the local file cache", d)
+			}
+
+			rn := digest.NewResourceName(d, remoteInstanceName, rspb.CacheType_CAS, repb.DigestFunction_BLAKE3)
+			rn.SetCompressor(repb.Compressor_ZSTD)
+			file, err := os.Open(path)
+			defer file.Close()
+			if err != nil {
+				log.Fatalf("Error opening file %s: %s", path, err)
+			}
+			_, err = cachetools.UploadFromReader(ctx, env.GetByteStreamClient(), rn, file)
+			if err != nil {
+				log.Fatalf("Error uploading file to remote cache %s: %s", path, err)
+			}
+		}
+
+		upload()
+	}
+}
+
+func uploadManifestRemoteCache(ctx context.Context, env environment.Env, key *fcpb.SnapshotKey, localACResult *repb.ActionResult, remoteInstanceName string) {
+	remoteManifestKey, err := snaploader.RemoteManifestKey(key)
+	if err != nil {
+		log.Fatalf("Error generating digest for snapshot key: %s", err)
+	}
+	acDigest := digest.NewResourceName(remoteManifestKey, remoteInstanceName, rspb.CacheType_AC, repb.DigestFunction_BLAKE3)
+	if err := cachetools.UploadActionResult(ctx, env.GetActionCacheClient(), acDigest, localACResult); err != nil {
+		log.Fatalf("Error uploading manifest to remote cache: %s", err)
```

Comment:
(no need to change this PR, just a small note for the future) log.Fatal prevents deferred cleanup funcs from being run (e.g. the temp dir cleanup) since it causes the process to immediately exit. As a workaround I like to use this boilerplate for Go binaries:

```
func main() {
  if err := run(); err != nil {
    log.Fatal(err.Error())
  }
}

func run() error {
  // ...
  if err != nil {
    return err
  }
  // ...
}
```